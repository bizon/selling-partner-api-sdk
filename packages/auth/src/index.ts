export * from './security-token-service'
export * from './access-token'
export * from './selling-partner-api-auth'
export * from './error'
