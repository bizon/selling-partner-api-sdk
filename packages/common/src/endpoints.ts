interface Endpoints {
  production: Record<string, string>;
  sandbox: Record<string, string>;
}

export const endpoints: Endpoints = {
  production: {
    'us-east-1': 'https://sellingpartnerapi-na.amazon.com',
    na: 'https://sellingpartnerapi-na.amazon.com',

    'eu-west-1': 'https://sellingpartnerapi-eu.amazon.com',
    eu: 'https://sellingpartnerapi-eu.amazon.com',

    'us-west-2': 'https://sellingpartnerapi-fe.amazon.com',
    fe: 'https://sellingpartnerapi-fe.amazon.com',
  },
  sandbox: {
    'us-east-1': 'https://sandbox.sellingpartnerapi-na.amazon.com',
    na: 'https://sandbox.sellingpartnerapi-na.amazon.com',

    'eu-west-1': 'https://sandbox.sellingpartnerapi-eu.amazon.com',
    eu: 'https://sandbox.sellingpartnerapi-eu.amazon.com',

    'us-west-2': 'https://sandbox.sellingpartnerapi-fe.amazon.com',
    fe: 'https://sandbox.sellingpartnerapi-fe.amazon.com',
  },
}

export const awsRegionByCode: Record<string, string> = {
  eu: 'eu-west-1',
  fe: 'us-west-2',
  na: 'us-east-1',
}
