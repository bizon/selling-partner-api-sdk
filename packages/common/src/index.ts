export * from './endpoints'
export * from './axios'
export * from './selling-partner-api-error'
