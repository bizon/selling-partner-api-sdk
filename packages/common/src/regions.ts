export type SellingPartnerRegion = 'na' | 'eu' | 'fe'

interface RegionConfiguration {
  awsRegion: string;
  endpoints: {
    production: string;
    sandbox: string;
  };
}

export const sellingPartnerRegions: Record<SellingPartnerRegion, RegionConfiguration> = {
  na: {
    awsRegion: 'us-east-1',
    endpoints: {
      production: 'https://sellingpartnerapi-na.amazon.com',
      sandbox: 'https://sandbox.sellingpartnerapi-na.amazon.com',
    },
  },
  eu: {
    awsRegion: 'eu-west-1',
    endpoints: {
      production: 'https://sellingpartnerapi-eu.amazon.com',
      sandbox: 'https://sandbox.sellingpartnerapi-eu.amazon.com',
    },
  },
  fe: {
    awsRegion: 'us-west-2',
    endpoints: {
      production: 'https://sellingpartnerapi-fe.amazon.com',
      sandbox: 'https://sandbox.sellingpartnerapi-fe.amazon.com',
    },
  },
}
