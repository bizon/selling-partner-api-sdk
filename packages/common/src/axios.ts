import axios, {type Method} from 'axios'
import {aws4Interceptor} from 'aws4-axios'
import axiosRetry from 'axios-retry'
import {requestLogger, responseLogger} from 'axios-logger'
import {sync as readPackageJson} from 'read-pkg-up'

import type {SellingPartnerApiAuth} from '@sp-api-sdk/auth'

import {SellingPartnerApiError} from './errors'
import {sellingPartnerRegions, type SellingPartnerRegion} from './regions'

const {packageJson} = readPackageJson()!

type RequestLogConfig = Exclude<Parameters<typeof requestLogger>[1], undefined>
type ResponseLogConfig = Exclude<Parameters<typeof responseLogger>[1], undefined>

export interface RateLimit {
  urlRegex: RegExp
  rate: number
  burst: number
  method: Method
}

export interface OnRetryParameters {
  delay: number
  rateLimit?: number
}

export interface ClientConfiguration {
  auth: SellingPartnerApiAuth
  region: SellingPartnerRegion
  userAgent?: string
  sandbox?: boolean
  rateLimiting?: {
    retry: boolean
    onRetry?: (retryInfo: OnRetryParameters) => void
  }
  logging?: {
    request?: RequestLogConfig | true
    response?: ResponseLogConfig | true
  }
}

export function createAxiosInstance(
  {
    auth,
    region,
    userAgent = `${packageJson.name}/${packageJson.version}`,
    sandbox = false,
    rateLimiting,
    logging,
  }: ClientConfiguration,
  rateLimits: RateLimit[],
) {
  const regionConfiguration = sellingPartnerRegions[region]
  if (!regionConfiguration) {
    throw new TypeError(`Unknown or unsupported region: ${region}`)
  }

  const instance = axios.create({
    headers: {
      'user-agent': userAgent,
    },
  })

  const endpoint = regionConfiguration.endpoints[sandbox ? 'sandbox' : 'production']

  if (rateLimiting?.retry) {
    axiosRetry(instance, {
      retryCondition: (error) => (error.response ? error.response.status === 429 : false),
      retryDelay(retryCount, error) {
        const amznRateLimit = error.response?.headers['x-amzn-ratelimit-limit']
        const url = new URL(error.config.url!)
        const rateLimit = amznRateLimit
          ? Number.parseFloat(amznRateLimit)
          : rateLimits.find(
              (rateLimit) =>
                rateLimit.method.toLowerCase() === error.config.method?.toLowerCase() &&
                rateLimit.urlRegex.exec(url.pathname),
            )?.rate
        const delay =
          rateLimit && !Number.isNaN(rateLimit) ? (1 / rateLimit) * 1000 + 1500 : 60 * 1000

        if (rateLimiting?.onRetry) {
          rateLimiting.onRetry({delay, rateLimit})
        }

        return delay
      },
    })
  }

  instance.interceptors.request.use(async (config) => {
    if (!config.headers) {
      config.headers = {}
    }

    config.headers['x-amz-access-token'] = await auth.accessToken.get()

    return config
  })

  instance.interceptors.request.use(async (config) => {
    const credentials = await auth.getCredentials()

    if (!credentials) {
      return config
    }

    return aws4Interceptor(
      {
        region: regionConfiguration.awsRegion,
        service: 'execute-api',
      },
      {
        accessKeyId: credentials.AccessKeyId ?? '',
        secretAccessKey: credentials.SecretAccessKey ?? '',
        sessionToken: credentials.SessionToken,
      },
    )(config)
  })

  instance.interceptors.response.use(
    async (response) => response,
    async (error: unknown) => {
      throw axios.isAxiosError(error) ? new SellingPartnerApiError(error) : error
    },
  )

  if (logging?.request) {
    const requestLoggerOptions = logging.request === true ? undefined : logging.request

    if (requestLoggerOptions?.headers) {
      console.warn(
        'WARNING: You have enabled logging of request headers, this can leak authentication information, you should disable in production.',
      )
    }

    instance.interceptors.request.use((config) =>
      requestLogger(config, {
        prefixText: `sp-api-sdk/${region}`,
        dateFormat: 'isoDateTime',
        method: true,
        url: true,
        params: false,
        data: true,
        headers: false,
        ...requestLoggerOptions,
      }),
    )
  }

  if (logging?.response) {
    const responseLoggerOptions = logging.response === true ? undefined : logging.response

    instance.interceptors.response.use((response) =>
      responseLogger(response, {
        prefixText: `sp-api-sdk/${region}`,
        dateFormat: 'isoDateTime',
        status: true,
        statusText: false,
        params: false,
        data: false,
        headers: true,
        ...responseLoggerOptions,
      }),
    )
  }

  return {axios: instance, endpoint}
}
