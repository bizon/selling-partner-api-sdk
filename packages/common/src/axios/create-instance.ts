import axios, {Method} from 'axios'
import {aws4Interceptor} from 'aws4-axios'
import axiosRetry from 'axios-retry'

import {SellingPartnerApiAuth} from '@sp-api-sdk/auth'

import pkg from '../../package.json'

export interface RateLimit {
  urlRegex: RegExp;
  rate: number;
  burst: number;
  method: Method;
}

export interface OnRetryParameters {
  delay: number;
  rateLimit?: number;
}

export type onRetry = (retryInfo: OnRetryParameters) => void

export interface ClientConfiguration {
  auth: SellingPartnerApiAuth;
  region: string;
  userAgent?: string;
  sandbox?: boolean;
  rateLimits?: RateLimit[];
  onRetry?: onRetry;
}

export function createAxiosInstance({
  auth,
  userAgent = `${pkg.name}/${pkg.version}`,
  region,
  rateLimits,
  onRetry
}: ClientConfiguration) {
  const instance = axios.create({
    headers: {
      'user-agent': userAgent
    }
  })

  if (rateLimits) {
    axiosRetry(instance, {
      retryCondition: error => error.response ? error.response.status === 429 : false,
      retryDelay: (retryCount, error) => {
        const amznRateLimit: string = error.response?.headers['x-amzn-ratelimit-limit']
        const url = new URL(error.config.url!)
        const rateLimit = amznRateLimit ? Number.parseFloat(amznRateLimit) : rateLimits.find(rateLimit => rateLimit.method.toLowerCase() === error.config.method?.toLowerCase() && rateLimit.urlRegex.exec(url.pathname))?.rate
        const delay = rateLimit && !Number.isNaN(rateLimit) ? (1 / rateLimit * 1000) + 1500 : (60 * 1000)

        if (onRetry) {
          onRetry({delay, rateLimit})
        }

        return delay
      }
    })
  }

  instance.interceptors.request.use(
    async config => {
      config.headers['x-amz-access-token'] = await auth.accessToken.get()

      return config
    }
  )

  instance.interceptors.request.use(async config => {
    const credentials = await auth.getCredentials()

    if (!credentials) {
      return config
    }

    return aws4Interceptor(
      {
        region,
        service: 'execute-api'
      },
      {
        accessKeyId: credentials.AccessKeyId ?? '',
        secretAccessKey: credentials.SecretAccessKey ?? '',
        sessionToken: credentials.SessionToken
      }
    )(config)
  })

  return instance
}
