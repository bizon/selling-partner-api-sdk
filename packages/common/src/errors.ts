import {URL} from 'url'

import {AxiosError} from 'axios'

export class SellingPartnerApiError<T = unknown, D = any> extends AxiosError<T, D> {
  public readonly innerMessage: string
  public readonly apiName?: string
  public readonly apiVersion?: string

  constructor(error: AxiosError<T, D>) {
    super('Unknown error', error.code, error.config, error.request, error.response)

    this.innerMessage = error.message

    if (error.config?.url) {
      const [apiName, apiVersion] = new URL(error.config.url).pathname.split('/').slice(1)
      const apiPrefix = `${apiName} (${apiVersion})`

      this.apiName = apiName
      this.apiVersion = apiVersion
      this.message = error.response
        ? `${apiPrefix} error: Response code ${error.response.status}`
        : `${apiPrefix} error: No response`
    }

    this.name = this.constructor.name
  }
}
