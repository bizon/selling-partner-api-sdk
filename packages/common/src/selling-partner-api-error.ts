import {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios'
import {URL} from 'url'

export class SellingPartnerApiError<T = any> extends Error implements AxiosError {
  public config: AxiosRequestConfig
  public code?: string
  public request?: any
  public response?: AxiosResponse<T>
  public isAxiosError: boolean

  public apiName?: string
  public apiVersion?: string

  private readonly axiosError: AxiosError

  constructor(error: AxiosError) {
    super()

    this.axiosError = error

    this.config = error.config
    this.code = error.code
    this.request = error.request
    this.response = error.response
    this.isAxiosError = error.isAxiosError

    const url = error.config.url ? new URL(error.config.url) : null

    if (url) {
      const [apiName, apiVersion] = url.pathname.split('/').slice(1)
      this.apiName = apiName
      this.apiVersion = apiVersion
    }
  }

  get name() {
    return this.constructor.name
  }

  get message() {
    if (this.apiName && this.apiVersion) {
      const apiPrefix = `${this.apiName} (${this.apiVersion})`

      if (!this.response) {
        return `${apiPrefix} error: No response`
      }

      return `${apiPrefix} error: Response code ${this.response.status}`
    }

    return 'Unknown error'
  }

  toJSON() {
    return this.axiosError.toJSON()
  }
}
