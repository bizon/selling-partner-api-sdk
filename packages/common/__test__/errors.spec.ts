import {AxiosError} from 'axios'

import {SellingPartnerApiError} from '../src/errors'

describe('src/errors', () => {
  const testError = new SellingPartnerApiError(new AxiosError('Original message', 'Code', {}))

  it('should expose the original axios error message', () => {
    expect(testError.innerMessage).toBe('Original message')
  })

  it('should be an instance of Error', () => {
    expect(testError).toBeInstanceOf(Error)
  })

  it('should be an instance of AxiosError', () => {
    expect(testError).toBeInstanceOf(AxiosError)
  })

  it('should be an instance of SellingPartnerApiError', () => {
    expect(testError).toBeInstanceOf(SellingPartnerApiError)
  })

  it('should extract the apiName and apiVersion from the config URL', () => {
    const error = new SellingPartnerApiError(
      new AxiosError('Message', 'Code', {
        url: 'https://domain/api-name/api-version?param=foo',
      }),
    )

    expect(error.apiName).toBe('api-name')
    expect(error.apiVersion).toBe('api-version')
  })

  it('should extract the status code from the axios response', () => {
    const config = {
      url: 'https://domain/api-name/api-version?param=foo',
    }

    const error = new SellingPartnerApiError(
      new AxiosError(
        'Message',
        'Code',
        config,
        {},
        {
          data: null,
          status: 404,
          statusText: 'Not found',
          config,
          headers: {},
        },
      ),
    )

    expect(error.message).toBe('api-name (api-version) error: Response code 404')
    expect(error.toJSON()).toMatchSnapshot()
  })

  it('should handle response-less errors', () => {
    const error = new SellingPartnerApiError(
      new AxiosError(
        'Message',
        'Code',
        {
          url: 'https://domain/api-name/api-version?param=foo',
        },
        {},
      ),
    )

    expect(error.message).toBe('api-name (api-version) error: No response')
    expect(error.toJSON()).toMatchSnapshot()
  })
})
