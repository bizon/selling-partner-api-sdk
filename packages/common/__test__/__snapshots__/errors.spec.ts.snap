// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/errors should extract the status code from the axios response 1`] = `
{
  "code": "Code",
  "columnNumber": undefined,
  "config": {
    "url": "https://domain/api-name/api-version?param=foo",
  },
  "description": undefined,
  "fileName": undefined,
  "lineNumber": undefined,
  "message": "api-name (api-version) error: Response code 404",
  "name": "SellingPartnerApiError",
  "number": undefined,
  "stack": undefined,
  "status": 404,
}
`;

exports[`src/errors should handle response-less errors 1`] = `
{
  "code": "Code",
  "columnNumber": undefined,
  "config": {
    "url": "https://domain/api-name/api-version?param=foo",
  },
  "description": undefined,
  "fileName": undefined,
  "lineNumber": undefined,
  "message": "api-name (api-version) error: No response",
  "name": "SellingPartnerApiError",
  "number": undefined,
  "stack": undefined,
  "status": null,
}
`;
